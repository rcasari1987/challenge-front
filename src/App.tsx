import React, { useEffect, useState } from "react";
import Box from "@mui/material/Box";
import Grid from "@mui/material/Grid";
import Modal from "@mui/material/Modal";
import { tvServices } from "./services";
import { Show } from "./types/tv-show";
import { MediaCard } from "./components";

function App() {
  const [tvShows, setTvShows] = useState<Show[]>();
  const [tvShow, setTvShow] = useState<Show>();
  const [open, setOpen] = React.useState(false);
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    const getTvShow = async () => {
      setIsLoading(true);
      try {
        const { data: response } = await tvServices.getShows();

        setTvShows(response.map((item) => item.show));
      } catch (error) {
      } finally {
        setIsLoading(false);
      }
    };

    getTvShow();
  }, []);

  const handleOpen = (id: number) => {
    const show = tvShows?.find((item) => item.id === id);
    setTvShow(show);
    setOpen(true);
  };
  const handleClose = () => {
    setTvShow(undefined);
    setOpen(false);
  };

  return (
    <Box
      sx={{ flexGrow: 1 }}
      style={{ backgroundColor: "#C3C3C3", padding: 20 }}
    >
      <Grid container spacing={2} data-testid="list">
        {tvShows &&
          tvShows.length &&
          tvShows.map((show) => (
            <Grid item xs={2} key={show?.id}>
              <MediaCard
                title={show?.name}
                src={show?.image?.medium}
                alt={`thumbnail-${show?.id}`}
                onMore={() => handleOpen(show?.id)}
              />
            </Grid>
          ))}
      </Grid>

      <Modal open={open} onClose={handleClose}>
        <MediaCard
          title={tvShow?.name!}
          description={tvShow?.summary}
          src={tvShow?.image?.medium}
          alt={`thumbnail-${tvShow?.id}`}
          style={style}
        />
      </Modal>
    </Box>
  );
}

const style = {
  position: "absolute" as "absolute",
  top: "50%",
  left: "50%",
  transform: "translate(-50%, -50%)",
  width: 500,
  boxShadow: 24,
};

export default App;
