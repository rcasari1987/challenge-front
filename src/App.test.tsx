import React from "react";
import renderer from "react-test-renderer";
import axios from "axios";

import App from "./App";
import { render, waitFor } from "@testing-library/react";

jest.mock("axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

beforeEach(() => {
  mockedAxios.get.mockClear();
});

test("render on api response items", async () => {
  mockedAxios.get.mockResolvedValue({
    data: [
      {
        score: 0.9079871,
        show: {
          id: 139,
          url: "https://www.tvmaze.com/shows/139/girls",
          name: "Girls",
          type: "Scripted",
          language: "English",
          genres: ["Drama", "Romance"],
          status: "Ended",
          image: {
            medium:
              "https://static.tvmaze.com/uploads/images/medium_portrait/31/78286.jpg",
            original:
              "https://static.tvmaze.com/uploads/images/original_untouched/31/78286.jpg",
          },
          summary:
            "<p>This Emmy winning series is a comic look at the assorted humiliations and rare triumphs of a group of girls in their 20s.</p>",
        },
      },
    ],
  });
  const { getByTestId, asFragment } = render(<App />);
  const listNode = await waitFor(() => getByTestId("list"));
  expect(listNode.children).toHaveLength(1);
  expect(asFragment).toMatchSnapshot();
});

test("render on api dont response items", async () => {
  mockedAxios.get.mockResolvedValue({ data: [] });
  const { getByTestId, asFragment } = render(<App />);
  const listNode = await waitFor(() => getByTestId("list"));
  expect(listNode.children).toHaveLength(0);
  expect(asFragment).toMatchSnapshot();
});
