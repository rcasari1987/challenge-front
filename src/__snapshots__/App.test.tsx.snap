// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render on api dont response items 1`] = `[Function]`;

exports[`render on api response items 1`] = `[Function]`;
